storage/merkle/proof_test.go

```diff
@@ -0,0 +1,166 @@
+package merkle
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestProofWithASingleKey tests proof generation and verification
+// when trie includes only a single value
+func TestProofWithASingleKey(t *testing.T) {
+
+	keyLength := 32
+	tree1, err := NewTree(keyLength)
+	assert.NoError(t, err)
+
+	key, val := randomKeyValuePair(32, 128)
+
+	replaced, err := tree1.Put(key, val)
+	assert.NoError(t, err)
+	require.False(t, replaced)
+
+	// work for an existing key
+	proof, existed := tree1.Prove(key)
+	require.True(t, existed)
+
+	isValid, err := proof.Verify(tree1.Hash())
+	assert.NoError(t, err)
+	require.True(t, isValid)
+
+	// fail for non-existing key
+	key2, _ := randomKeyValuePair(32, 128)
+
+	proof, existed = tree1.Prove(key2)
+	require.False(t, existed)
+	require.Nil(t, proof)
+}
+
+// TestValididateFormat tests cases a proof can not be valid
+func TestValididateFormat(t *testing.T) {
+
+	// construct a valid proof
+	keyLength := 32
+	key := make([]byte, keyLength)
+	key[0] = uint8(5)
+	value := make([]byte, 128)
+	value[0] = uint8(6)
+
+	key2 := make([]byte, keyLength)
+	key2[0] = uint8(4)
+	value2 := make([]byte, 128)
+
+	tree1, err := NewTree(keyLength)
+	assert.NoError(t, err)
+	replaced, err := tree1.Put(key, value)
+	assert.NoError(t, err)
+	require.False(t, replaced)
+	replaced, err = tree1.Put(key2, value2)
+	assert.NoError(t, err)
+	require.False(t, replaced)
+	proof, existed := tree1.Prove(key)
+	require.True(t, existed)
+
+	// invalid key size
+	proof.Key = make([]byte, 0)
+	err = proof.valididateFormat()
+	assert.Error(t, err)
```

Comment:
the primary criterium here to test for is that the expected error _type_ is returned, because calling code should work with the type and not rely on parsing the error message:
```suggestion
	assert.True(t, IsMalformedProofError(err))
```

Reply:
updated